### ISSUE TITLE
Specify type for variable or allow parameter allow_input_downcast=True for theano function

### ISSUE THREAD

**Issue by rflamary** [pymanopt/pymanopt#32]
Hello,

Thank you for all the work! I've been using your toolbox for a pretty complex optimization problem and it's very easy to use. I will of course cite you as soon as anything is published.

I ran into a small problem when using a theano loss. I have access to CUDA cards and so I use the GPU configuration of theano (with floatX=float32) with following .theanorc  

```
[global]
floatX = float32
device = gpu
```

So I converted all my data to float32 and ran the solver. I got a standard theano type error that shouldn't be here since all the data is converted. I think that this comes from the  the initial value that is set in in pymanopt . The default numpy array is in float64 and it probably leads to problem when computing the loss with the theano function.

I simply edited 
https://github.com/pymanopt/pymanopt/blob/master/pymanopt/tools/autodiff/_theano.py#L45
and added the parameter  "allow_input_downcast=True" to the function and the problem disappeared.

Another approach would be to allow pymanopt to specify the type of the initial value.

Since both the propositions will have a strong impact on the API (add a new parameter to main classes) I will let you decide which and how to implement it.

Note that I don't really need implemented since everything works for me now but I believe this kind of sneaky problems can easily happen to other people (I have a pretty standard theano configuration).

**Comment by j-towns** [comment]
Thanks for raising this. I agree, we should really support float32 accross the whole toolbox without any hassle for the user. We'll have a look and try to work out what changes we need to make, hopefully shouldn't be too difficult.

**Comment by nkoep** [comment]
I'm closing this for now since it's no longer relevant after dropping theano support in #155. This might still be relevant for the pytorch and tensorflow backends with support for GPUs, in which case we will re-open the ticket once we encounter similar problems.
